### PR TITLE
feat(android, sdk12)!: update to facebook-android-sdk 12

### LINKED ISSUE #102: CI is a little flaky, needs resiliency added to network activity
## 🐛 Bug Report

CI fails sometimes, and it is not the module's fault.

Specifically sometimes yarn fails to install packages, though I imagine we will see Pod install failures in the future as well.

## To Reproduce

Run CI a few times, and examine the results.

## Expected Behavior

CI does not fail on transient network failures

## Code Example

A typical solution would be to add https://github.com/nick-invision/retry/ to the workflows, as in other repos.

An example:

https://github.com/invertase/react-native-firebase/blob/47d226526fec4dcee6edc5f5c2e37f430e91cae9/.github/workflows/tests_e2e_ios.yml#L78-L84

```yaml
      - name: Yarn Install
        uses: nick-invision/retry@v2
        with:
          timeout_minutes: 10
          retry_wait_seconds: 60
          max_attempts: 3
          command: yarn --no-audit --prefer-offline
```

## Environment

Github actions runner

### LINKED ISSUE #115: Move from jcenter to maven
Is there any plan to remove usages of jcenter() repository from your build scripts and migrate your build to other maven repositories?

### PR DESCRIPTION
**Pull request by mikehardy**
This PR should be reviewed commit by commit and rebase-merged when ready. I'll keep the commits clean if they need changes, and re-push

Each commit has it's own message and is separate, for easy review

The big idea is:

- "update all the things", that is: the github workflows, the android SDK, the example app, the javascript dependencies, and then prove android SDK 31 works

Fixes #115 
Related #55 (takes care of the android part)
Fixes #102

Test Plan:

I cleanly installed and ran example app on android emulator and ios simulator repeatedly while working through the upgrades, everything worked

### CODE REVIEW

**Comment by mikehardy** [comment]
@thebergamo I'll let this sit for a day or two but it's all green in CI and passes locally so plan is to merge then release a breaking change release since an API was removed (both from iOS and Android SDKs, so removed here as well). It's a major advance for the library though to move forward on the underlying SDKs and update the example so I'm excited to get it out there

**Comment by thebergamo** [comment]
I never seem so well documented PR!
Kudos ❤️
There is just a typo in the example besides it, feel free to move forward

File: RNFBSDKExample/README.md [hunk 1]
```diff
@@ -4,6 +4,8 @@ This example is extremely basic but hopefully still useful.
 
 You run it by cloning the main repo, and running `yarn` then `yarn example:android` or `yarn example:ios`
 
+Note: If you hae run it before, and are trying to update it `rm yarn.lock node_modules ios/Podfile.lock && yarn && cd ios && pod install` in order to make sure you have fresh code, or if you are developing run `yarn example:copy` from the parent direcotry. Why? The example references the module code via a `../` reference, and the and it may contain stale files unless you clear the lockfiles/modules and reinstall.
```
Comment (thebergamo) [inline RNFBSDKExample/README.md:7]: Just a typo here :)
What about having this command as an npm script also?

**Review by thebergamo — APPROVED** [review]

File: RNFBSDKExample/README.md (hunk 1 above)
Comment (mikehardy) [reply RNFBSDKExample/README.md:7]: Argh typos, will fix :-)

I did put `yarn example:copy` in the run scripts, I could put a `yarn example:clean` as well to package up the other ones, good idea. I'll add that and fix the typo and repush when I get a minute

Thanks for the review! With regard to documenting etc, I'm always trying to code for the "me" in the future (like, just a month or two for now) that will have forgotten all this and will need the information for some future bug (there's always a future bug...). Then the documentation and commits and comments become a valuable gift from Me-from-the-past :nerd_face:

**Comment by hannojg** [comment]
(just wanted to add that we are using this PR also for our app and as far as I can tell everything works 👍 Thank you very much for the work put in here!)

**Comment by mikehardy** [comment]
Thanks very much for the success report! I'm just tidying up the example app generator right now but I can't see anything to change for the *actual* android forward port (it was really trivial anyway, thankfully) so when this actually merges+releases it won't really be a change for you to switch back to the release version

Cheers

**Comment by mikehardy** [comment]
Okay finally got this to go 100% green - nothing substantial in the module code has changed.
The main changes in the last day were 100% around getting the example app to configure itself more thoroughly (the underlying SDKs added a lot of configuration elements...) and getting the CI to more thoroughly exercise things while getting the developer experience / documentation lined up all good to go now - I need to release this so I can move on to some pressing work deadlines :crossed_fingers: - if something breaks I'll be listening though

**Comment by mikehardy** [comment]
@thebergamo I think we're ready for a release here but I wanted to coordinate with you to see if you have any thoughts just in case. It should be a major version, and I *think* this commit message is exactly what is needed to have that work automatically:

https://github.com/thebergamo/react-native-fbsdk-next/commit/cc1946380f1164b0e78f300a99db7644c7690b91

The `!` is supposed to be what triggers it, and then the `BREAKING CHANGE` notes are supposed to be auto-populated in the changelog / release notes? That's how it is supposed to work, and it has worked for me here:

- Log: https://github.com/invertase/react-native-firebase/blob/main/CHANGELOG.md#breaking-changes
- Commit: https://github.com/invertase/react-native-firebase/commit/f60afe158b2dc823bd7169e36c3e428470576c7e

What do you think? Ready for the release button? :-)

**Comment by thebergamo** [comment]
I don't see any issues with it and I agree with you, it should be the fine as per the commit message.

Go for it tiger 🐯

**Comment by mikehardy** [comment]
Great :rocket:  :crossed_fingers:

**Comment by mikehardy** [comment]
Great CI setup, it all worked perfectly! https://github.com/thebergamo/react-native-fbsdk-next/releases/tag/v5.0.0
Developers will know exactly how to handle it if it affects them

**Pull request merged**